_projects/standardised-file-format.md

```diff
@@ -0,0 +1,70 @@
+collaborating_projects:
+  - coala
+desc: "Provide support to a standardised file format(TOML) for writing configurations"
+difficulty: hard
+initiatives:
+  - GSoC
+markdown: standardised-file-format.md
+mentors:
+  - 
+name: "Support TOML for writing configurations"
+requirements:
+  - "The applicant should have at least one patch merged to
+    coala"
+  - "The applicant should be familiar with Python."
+  - "The applicant should be familiar with coala configuration system"
+tags:
+  - Core
+  - Configuration
+  - TOML
+---
+
+This project aims to support a standardised file format like TOML for writing settings.
+Currently coala has a INI style config. It has `LineParser` and `ConfParser` classes to
+parse the settings into sections. These sections are later executed by running the
+appropriate bear. coala also a `ConfWriter` class to write sections back into settings
+file like `.coafile`
+
+This approach has certain limitations. Custom sub-level parsing in INI style config
+is tough and error prone. It makes `LineParser` and `ConfParser` more complex and 
+bug prone and also would slow down the speed of development. A solution would be to 
+use standardised file format like TOML. Using `toml.load` on the config file will 
+return a dictionary. This makes custom sub-level parsing not only easy and accurate
+but also less error prone. Also more importantly completely removes the need 
+for `LineParser`.
+
+TOML provides flexibility, consistency, backwards compatibility and standard. It has
+widespread acceptance, documentation and robust parsers. TOML has been recommended as
+the configuration file for software packages to specify their build dependencies. Due 
+to its popularity it offers a reduced learning curve compared to INI style config.
+
+#### Milestones
+
+##### PREPARATION/BONDING
+
+- The applicant has created and merged a cEP for this project.
+- The applicant should be familiar with the codebase.
+- The applicant should know how to work with the community.
+
+##### CODING PHASE 1
+
+- Create a class named `TomlConfParser` to load settings from TOML
+  file into sections.
+- The corresponding code is properly tested, documented, reviewed,
+  and merged.
+- Write and test all possible settings in the new parser. 
+
+##### CODING PHASE 2
+
+- Create a class named `TomlConfWriter` to write sections into  
+  TOML file.
+- The corresponding code is properly tested, documented, reviewed,
+  and merged.
+- Test all possible settings in the new writer.
+
+##### CODING PHASE 3
+
+- Support required and missing features in the parser by making PR's to
+  [toml](https://github.com/uiri/toml)
+- New coala tutorials and user guides have been created to show how to
```

Comment:
IMO it would be great to have a tool which converts existing .coafile into TOML config file.

Reply:
Thanks 😊
Regarding the first question , https://www.python.org/dev/peps/pep-0518/#overview-of-file-formats-considered explains why TOML is better compared to all other formats. So TOML should be able to do much more that INI or other available config formats. 

`converts existing .coafile into TOML config file` is a great idea. It should be possible to implement with this project.  Once `TomlConfWriter` is implemented, the sections loaded from `.coafile` can be passed into it to generate a TOML config file.

Reply:
can you also add this idea too? Also check why travis fails.

Reply:
I will add it.

Reply:
@PrajwalM2212 regarding question-
`is there any limitation with adopting TOML? I don't know why people built a custom parser and configuration format instead of using a standard one, but maybe there is some reason.
`
I know TOML format is better but there will be a valid reason imo that they not started with TOML format from the start.

There will be many edge cases in integrating the TOML format, you should write down all the cases one by one and also write down what will break, how you are going to solve it and what will improve.

Ping @jayvdb @Makman2

Reply:
@RaiVaibhav 
1. When coala started out, the syntax did not require custom sub level parsing, so INI style config was good .  But when we started to implement aspects we had to face with tough task of custom sub level parsing. This makes `LineParser` and `ConfParser` more complex and error prone.
2.  `There will be many edge cases in integrating the TOML format` .  
The only challenge is to define the mapping between current coafile syntax and TOML syntax. 
`bears = SpaceConsistencyBear, LineLengthBear` can be easily mapped to `bears = [ SpaceConsistencyBear, LineLengthBear ]`.  
But what does what does files += app.py  map into in TOML ? This has to be decided by the community. What syntax they like. No matter what syntax they may like, it can be easily implemented in TOML. 
Other challenge is that TOML parser does not load comments. There is an existing issue to solve this. 
3. `write down what will break`
According to my current plan, absolutely nothing shall break down. I shall not be altering any functionality.  I will add classes like `TomlConfWriter` , `TomlConfParser`  etc, but not break anything. The plan is to give full support to TOML like coafile.  

I have written a very detailed plan in https://github.com/PrajwalM2212/Coala-GSoC-Plan. You can have a look at that.